File: 139. Word Break/Solution.java
class Solution {
    Set<Integer> memo = new HashSet<>();
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> d = new HashSet<>(wordDict);
        return r(s, d, 0);
    }

    public boolean r(String s, Set<String> d, int i) {
        if (i == s.length()) return true;
        if (memo.contains(i)) return false;
        for (int j = i + 1; j < s.length() + 1; j++) {
            if (d.contains(s.substring(i, j)) && r(s, d, j)) return true;
        }
        memo.add(i);
        return false;
    }
}

File: 139. Word Break/Solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        self.memo = set()
        return self.r(s, set(wordDict), 0)
    def r(self, s, d, i):
        if i == len(s): return True
        if i in self.memo: return False
        for j in range(i + 1, len(s) + 1):
            if s[i:j] in d and self.r(s, d, j): return True
        self.memo.add(i)
        return False
